### ISSUE TITLE
Update publications crawler to correct publications information; e.g. date

### ISSUE THREAD

**Issue by milen-sage** [Sage-Bionetworks/porTools#10]
- [x] Check with @avanlinden if publications [here](https://www.synapse.org/#!Synapse:syn20448807/tables/) are correct and obtain examples where publication entries have errors in publication information. 
- [x] Check out the publication update documentation [here] and take a look at the scripts linked in the docs, that currently pull publication information automatically. Try to reproduce the errors in publication info (https://sagebionetworks.jira.com/wiki/spaces/SBC/pages/2072903764/Publications+Script). 
- [x] Fix the script(s) and pull updated publication information. If the fix seems elusive, correct erroneous publication info following the directions in the docs above for manual update. Depending on the number of publications with errors, this may require a script that updates the synapse publication folder. This script could be ad-hoc.

This should position us well to tackle the more general and generic solution for publication crawling. In later sprints and issues, we will start outlining the requirements for a robust service that crawls the web (perhaps not limited to pubmed) for publications' information. That may rely on a commercial service like dimensions.ai.

In the process, we'll likely converge on publication information schema (e.g. along the lines of this [one](https://sagebionetworks.jira.com/wiki/spaces/PS/pages/820609093/AMP-AD+Publication+Schema). And start from there

Related, here is another [script](https://github.com/Sage-Bionetworks/syndccutils/tree/pubmed/python/syndccutils) that crawls similar publication information but in python. 

@ychae we can sync up to outline a user story map for the scenario of updating publication information.

Acceptance criteria: publication with erroneous entries are updated in AMP-AD resources.

**Comment by milen-sage** [comment]
@ychae it looks like we can't assign people to issues in this repo? Is this a repo access issue?

**Comment by avanlinden** [comment]
@milen-sage It must be an access issue -- I think if I give you write permission to the repo you can assign. Is there like an "IBC admin" github team I can give access to, or should I give it to you and @ychae individually?

**Comment by ychae** [comment]
@milen-sage and @avanlinden you're right, it's an access level issue. We haven't created a github team yet, but we should do that soon. @avanlinden can you give myself and Milen write level access to be able to assign in the meantime? I'll try to get the various teams created in the next week or so.

**Comment by avanlinden** [comment]
@ychae Done!

**Comment by lakikowolfe** [comment]
## Notes from meeting with Abby

### Overview
- `porTools` is a package that contains scripts to help with managing data upload to the amp AD portal.
- Script to update publication metadata (author, date, etc) does not reliably work. Can be difficult to get to run, outputs opaque errors due to nested function structure and `easyPubMedR` package, does not output the correct date for publications
- Script has never really worked well, should probably be considered a first draft

### update-publications-ad.R
- Pulls grant numbers from synapse
- queries pubmed with grant numbers
- Most failures occur within query functions
- Not sure where the date error occurs or why. Could be query, or maybe when the data is being parsed/cleaned.

### easyPubMed
- Not a great package to rely on. Maintained by one person, unhelpful error messages
- PubMed API is changing (has changed?) will this effect anything?
- Python has biopython entrez module? Or maybe other ways to query pubmed. Could use reticulate to take advantage of these packages.

### Other Sage tools
- There are a few different tools floating around Sage that scrape PubMed. Move towards one tool?
- Milen linked python pubcrawler above
- Is there any effort to create a single tool?

**Comment by lakikowolfe** [comment]
## Replicate Error

### Get data using `EasyPubMed`
```
# Get info on grants from syn17024229
grants <- syn$tableQuery("SELECT \"Grant Number\", grantSerialNumber, Program FROM syn17024229"
)$asDataFrame()

## grantSerialNumber becomes the query
## Remove rows that have NaN or NA or empty string for the serial number
grants <- grants[!(grants$grantSerialNumber %in% c(NaN, NA, "")), ]

# get pmids for a single grant ID
grant_info <- easyPubMed::get_pubmed_ids(grants$grantSerialNumber[[1]])

# pull pmids out for each pub under the grant serial number
grant_pmids <- easyPubMed::fetch_all_pubmed_ids(grant_info)
grant_pmids <- as.integer(grant_pmids)

# pick a single pmid (randomly chose the 50th pmid in the vector)
# use get_pub_med_ids again but only for the single pubmed id
# This is becuase fetch_pubmed_data() requires structured xml output from get_pubmed_ids()
pmid_structured_q <- easyPubMed::get_pubmed_ids(grant_pmids[50])

# pull down text blob for pmid
blob <- easyPubMed::fetch_pubmed_data(pmid_structured_q)

# blob to df
df <- easyPubMed::article_to_df(blob)
```

### Compare Results

- The `EasyPubMed` query results shows that the paper was published 2021-08-06
- [Manual PubMed query](https://pubmed.ncbi.nlm.nih.gov/32666270/) shows that paper was published electronically 2020-07-14 and a part of the 2020-09 publication of the journal.

It is not clear where `EasyPubMed` is getting this date?

**Comment by lakikowolfe** [comment]
What's left:
- [ ] Test script output
- [ ] Have @avanlinden run script/give feedback

**Comment by lakikowolfe** [comment]
@avanlinden I believe that I've successfully refactored the publication crawler to use `rentrez` instead of `easyPubMed`. This solved the date problem! 

I've held off on pushing the new dataset to Synapse because I thought you (or someone else) might want to try running the script and looking at the output first? Let me know what you're thinking!

**Comment by avanlinden** [comment]
@lakikowolfe This looks awesome! I'll run through it as soon as I get a chance just to check. Thank you!

**Comment by avanlinden** [comment]
@lakikowolfe Sorry for the delay on this -- I'm actually testing your branch right now and the script seems to have stalled out. Going to let it go for a little while longer and then do some more digging, but I'll update you as I go.

**Comment by lakikowolfe** [comment]
@milen-sage I agree. I went with the quick implementation for now.

I think we could add another parameter to the script where a user could input an API key if they have one. If they don't have one the script could default to the sleep time solution.

**Comment by lakikowolfe** [comment]
Notes on NCBI API Keys from rentrez tutorial:

>By default, the NCBI limits users to making only 3 requests per second (and rentrez enforces that limit). Users who register for an "API key" are able to make up to ten requests per second. Getting one of these keys is simple, you just need to [register for "my ncbi" account](https://www.ncbi.nlm.nih.gov/account/) then click on a button in the [account settings page](https://www.ncbi.nlm.nih.gov/account/settings/).

>Once you have an API key, rentrez will allow you to take advantage of it. For one-off cases, this is as simple as adding the api_key argument to given function call.

>If you use rentrez often you should edit your .Renviron file (see r help(Startup) for description of this file) to include your key. Doing so will mean all requests you send will take advantage of your API key.
`ENTREZ_KEY=ABCD123`

**Comment by avanlinden** [comment]
Next error: 500 character Synapse annotation limit. 

```
Getting publication PubMed IDs associated with provided grants
  |======================================================================| 100%


Querying PubMed for Publication Metadata
  |======================================================================| 100%Error in py_call_impl(callable, dots$args, dots$keywords) : 
  SynapseHTTPError: 400 Client Error: 
Total String value too long. Can be at most 500 characters.

Detailed traceback:
  File "/Users/alinden/Library/r-miniconda/envs/r-reticulate/lib/python3.7/site-packages/synapseclient/client.py", line 1155, in store
    annotations = self.set_annotations(Annotations(properties['id'], properties['etag'], annotations))
  File "/Users/alinden/Library/r-miniconda/envs/r-reticulate/lib/python3.7/site-packages/synapseclient/client.py", line 1467, in set_annotations
    body=json.dumps(synapseAnnos)))
  File "/Users/alinden/Library/r-miniconda/envs/r-reticulate/lib/python3.7/site-packages/synapseclient/client.py", line 3921, in restPUT
    response = self._rest_call('put', uri, body, endpoint, headers, retryPolicy, requests_session, **kwargs)
  File "/Users/alinden/Library/r-miniconda/envs/r-reticulate/lib/python3.7/site-packages/synapseclient/client.py", line 3871, in _rest_call
    self._handle_synapse_http_error(response)
  File "/Users/alinden
Calls: store_as_annotations -> <Anonymous> -> .f -> <Anonymous> -> py_call_impl
Execution halted
```

Should be easy to fix by trimming the pubmed query output to under 500 characters before annotations are set on the entities.

**Comment by lakikowolfe** [comment]
Ok [this](https://github.com/Sage-Bionetworks/porTools/commit/7dea1738a817f298ed9fc2ccb0b7dcb51a4910f0) should have solved the 500 char problem. It only really affects the author col but figured it's safer to run all the columns through it.

@avanlinden please give the script a run and let me know how it goes!

**Comment by lakikowolfe** [comment]
I did notice there are two entry rows that have possibly problematic NAs. 

1. Seems to be a problem with pubmed ID. I'm removing this row for now, but need to run down what is causing this in the long term.
```
pubmed_id pubdate title journal authors DOI   year  entity_name             grant       Program      
NA        NA      NA    NA      NA      NA    NA    NA NA NA (Pubmed ID NA) R56AG061837 Resilience-AD
```
2. [One publication doesn't have any authors listed](https://pubmed.ncbi.nlm.nih.gov/30905591/) and this shows up in the entity ID. I'm not sure if this matters?
```
# entity name
NA Alzheimer's and dementia - the journal of the Alzheimer's Association 2019 (Pubmed ID 30905591)
```
This is due to some error handling when pulling the authors. I can also have no author show up if I used an empty string instead of NA to replace missing authors. Let me know if you have a preferance!

**Comment by avanlinden** [comment]
@lakikowolfe This works! The last thing I ran into was yet another unaccounted for character from a manuscript title that Synapse will not allow as an entity name ("["), I made a quick change to catch those brackets before entity creation and after that it worked great!

My changes are on the `even-hackier-cleaning` branch, let me know if you want me to open a PR to add them into your branch or the master branch or what.
